src/Orleans.Runtime/Messaging/Gateway.cs

```diff
@@ -170,6 +170,19 @@ internal void RecordClosedSocket(Socket sock)
 
         internal SiloAddress TryToReroute(Message msg)
         {
+            // ** Special routing rule for system target here **
+            // When a client make a request/response to/from a SystemTarget, the TargetSilo can be set to either 
+            //  - the GatewayAddress of the target silo (for example, when the client want get the cluster typemap)
+            //  - the "internal" Silo-to-Silo address, if the client want to send a message to a specific SystemTarget
+            //    activation that is on a silo on which there is no gateway available (or if the client is not
+            //    connected to that gateway)
+            // So, if the TargetGrain is a SystemTarget we always trust the value from Message.TargetSilo and forward
+            // it to this address...
+            // EXCEPT if the value is equal to the current GatewayAdress: in this case we will return
+            // null and the local dispatcher will forward the Message to a local SystemTarget activation
+            if (msg.TargetGrain.IsSystemTarget && !msg.TargetSilo.Matches(this.gatewayAddress))
```

Comment:
Can either of these values ever be null here?

Reply:
It doesn't make sense to have a message without a `TargetGrain`. `TargetSilo` is always set on the client, but I think it is cheap enough to add a null check here to be more defensive, I will add it